model/storage-private/src/main/java/org/keycloak/storage/UserStorageManager.java

```diff
@@ -422,7 +422,11 @@ public UserModel getUserById(RealmModel realm, String id) {
     public UserModel getUserByUsername(RealmModel realm, String username) {
         UserModel user = localStorage().getUserByUsername(realm, username);
         if (user != null) {
-            return importValidation(realm, user);
+            user = importValidation(realm, user);
+            // Case when username was changed directly in the userStorage and doesn't correspond anymore to the username from local DB
+            if (user != null && username.equalsIgnoreCase(user.getUsername())) {
```

Comment:
Looks like we assume this check is done by https://www.keycloak.org/docs/26.2.5/server_development/#importeduservalidation-interface. 

The `importValidation` will return the same model instance if your provider is not implementing it. In this case, you will eventually return a stale model, potentially different than the username from the provider.

But I guess our recommendation is to implement the `ImportedUserValidation` interface for user storage providers if the underlying store is writable and you want consistency?